File: repos_estudiantes_ALVAREZ_segunda_entrega/AGUILAR_MACHICADO_NILS_REINER/PRACTICA 2/HERENCIA/Ejercicio3.py
from datetime import datetime

class Persona:
    def __init__(self, ci, nombre, apellido, celular, fecha_nac):
        self.ci = ci
        self.nombre = nombre
        self.apellido = apellido
        self.celular = celular
        self.fecha_nac = fecha_nac  # formato: "YYYY-MM-DD"

    def edad(self):
        nacimiento = datetime.strptime(self.fecha_nac, "%Y-%m-%d")
        hoy = datetime.now()
        return hoy.year - nacimiento.year - ((hoy.month, hoy.day) < (nacimiento.month, nacimiento.day))

    def mostrar(self):
        return f"{self.nombre} {self.apellido} ({self.ci}) - Cel: {self.celular}, Nac: {self.fecha_nac}"

class Estudiante(Persona):
    def __init__(self, ci, nombre, apellido, celular, fecha_nac, ru, fecha_ingreso, semestre):
        super().__init__(ci, nombre, apellido, celular, fecha_nac)
        self.ru = ru
        self.fecha_ingreso = fecha_ingreso
        self.semestre = semestre

    def mostrar(self):
        return super().mostrar() + f" | RU: {self.ru}, Ingreso: {self.fecha_ingreso}, Semestre: {self.semestre}"

class Docente(Persona):
    def __init__(self, ci, nombre, apellido, celular, fecha_nac, nit, profesion, especialidad, sexo):
        super().__init__(ci, nombre, apellido, celular, fecha_nac)
        self.nit = nit
        self.profesion = profesion
        self.especialidad = especialidad
        self.sexo = sexo

    def mostrar(self):
        return super().mostrar() + f" | NIT: {self.nit}, Profesion: {self.profesion}, Especialidad: {self.especialidad}, Sexo: {self.sexo}"

personas = [
    Estudiante("123", "Ana", "Perez", "7891234", "1997-05-10", "RU001", "2021-01-10", 5),
    Estudiante("456", "Luis", "Gomez", "7891235", "2006-09-15", "RU002", "2023-01-10", 2),
    Docente("789", "Carlos", "Perez", "7891236", "1980-02-20", "NIT001", "Ingeniero", "Sistemas", "M"),
    Docente("321", "Maria", "Gomez", "7891237", "1985-07-01", "NIT002", "Licenciada", "Matemática", "F")
]

print("--- Estudiantes mayores de 25 años ---")
for p in personas:
    if isinstance(p, Estudiante) and p.edad() > 25:
        print(p.mostrar())

print("\n--- Docente Ingeniero, masculino y mayor ---")
docentes_m = [p for p in personas if isinstance(p, Docente) and p.profesion == "Ingeniero" and p.sexo == "M"]
if docentes_m:
    mayor = max(docentes_m, key=lambda d: d.edad())
    print(mayor.mostrar())

print("\n--- Personas con mismo apellido ---")
for i in range(len(personas)):
    for j in range(i + 1, len(personas)):
        if personas[i].apellido == personas[j].apellido:
            print(personas[i].mostrar())
            print(personas[j].mostrar())

File: repos_estudiantes_ALVAREZ_segunda_entrega/AGUILAR_MACHICADO_NILS_REINER/PRACTICA 2/HERENCIA/Ejercicio3.java
import java.time.LocalDate;
import java.time.Period;

class Persona {
    private String ci, nombre, apellido, celular;
    private LocalDate fechaNac;

    public Persona(String ci, String nombre, String apellido, String celular, LocalDate fechaNac) {
        this.ci = ci;
        this.nombre = nombre;
        this.apellido = apellido;
        this.celular = celular;
        this.fechaNac = fechaNac;
    }

    public int getEdad() {
        return Period.between(fechaNac, LocalDate.now()).getYears();
    }

    public String getApellido() {
        return apellido;
    }

    public String toString() {
        return nombre + " " + apellido + " (" + ci + ") - Cel: " + celular + ", Nac: " + fechaNac;
    }
}

class Estudiante extends Persona {
    private String ru;
    private LocalDate fechaIngreso;
    private int semestre;

    public Estudiante(String ci, String nombre, String apellido, String celular, LocalDate fechaNac,
                      String ru, LocalDate fechaIngreso, int semestre) {
        super(ci, nombre, apellido, celular, fechaNac);
        this.ru = ru;
        this.fechaIngreso = fechaIngreso;
        this.semestre = semestre;
    }

    public String toString() {
        return super.toString() + " | RU: " + ru + ", Ingreso: " + fechaIngreso + ", Semestre: " + semestre;
    }
}

class Docente extends Persona {
    private String nit, profesion, especialidad, sexo;

    public Docente(String ci, String nombre, String apellido, String celular, LocalDate fechaNac,
                   String nit, String profesion, String especialidad, String sexo) {
        super(ci, nombre, apellido, celular, fechaNac);
        this.nit = nit;
        this.profesion = profesion;
        this.especialidad = especialidad;
        this.sexo = sexo;
    }

    public String getProfesion() {
        return profesion;
    }

    public String getSexo() {
        return sexo;
    }

    public String toString() {
        return super.toString() + " | NIT: " + nit + ", Profesion: " + profesion + ", Especialidad: " + especialidad + ", Sexo: " + sexo;
    }
}

class TestPersonas {
    public static void main(String[] args) {
        Persona[] personas = new Persona[] {
                new Estudiante("123", "Ana", "Perez", "7891234", LocalDate.of(1997, 5, 10), "RU001", LocalDate.of(2021, 1, 10), 5),
                new Estudiante("456", "Luis", "Gomez", "7891235", LocalDate.of(2006, 9, 15), "RU002", LocalDate.of(2023, 1, 10), 2),
                new Docente("789", "Carlos", "Perez", "7891236", LocalDate.of(1980, 2, 20), "NIT001", "Ingeniero", "Sistemas", "M"),
                new Docente("321", "Maria", "Gomez", "7891237", LocalDate.of(1985, 7, 1), "NIT002", "Licenciada", "Matemática", "F")
        };

        System.out.println("--- Estudiantes mayores de 25 años ---");
        for (Persona p : personas) {
            if (p instanceof Estudiante && p.getEdad() > 25) {
                System.out.println(p);
            }
        }

        System.out.println("\n--- Docente Ingeniero, masculino y mayor ---");
        Docente mayorIngeniero = null;
        for (Persona p : personas) {
            if (p instanceof Docente d && d.getProfesion().equals("Ingeniero") && d.getSexo().equals("M")) {
                if (mayorIngeniero == null || d.getEdad() > mayorIngeniero.getEdad()) {
                    mayorIngeniero = d;
                }
            }
        }
        if (mayorIngeniero != null) System.out.println(mayorIngeniero);

        System.out.println("\n--- Personas con mismo apellido ---");
        for (int i = 0; i < personas.length; i++) {
            for (int j = i + 1; j < personas.length; j++) {
                if (personas[i].getApellido().equals(personas[j].getApellido())) {
                    System.out.println(personas[i]);
                    System.out.println(personas[j]);
                }
            }
        }
    }
}